File: backend/src/controllers/policiesController.js
const { validationResult } = require('express-validator');
const Policy = require('../models/Policy');

const getPolicies = async (req, res, next) => {
  try {
    const { product_type_id, agent_id, date_from, date_to, search, page = 1, limit = 20 } = req.query;

    const filters = {};
    if (product_type_id) filters.product_type_id = parseInt(product_type_id);
    if (date_from) filters.date_from = date_from;
    if (date_to) filters.date_to = date_to;
    if (search) filters.search = search;

    const { role_name, user_id, branch_id } = req.user;

    if (role_name === 'Sales Agent') {
      filters.agent_id = user_id;
    } else if (agent_id) {
      filters.agent_id = parseInt(agent_id);
    }

    const pagination = { page: parseInt(page), limit: Math.min(parseInt(limit), 100) };
    const { data, total } = await Policy.findAll(filters, pagination);

    return res.status(200).json({
      success: true,
      data,
      pagination: {
        page: pagination.page,
        limit: pagination.limit,
        total,
        totalPages: Math.ceil(total / pagination.limit)
      }
    });
  } catch (err) {
    next(err);
  }
};

const createPolicy = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const { customer_name, policy_number, lead_id, product_type_id, premium, start_date, end_date, agent_id } = req.body;

    const policy = await Policy.create({
      customer_name,
      policy_number,
      lead_id,
      product_type_id,
      premium,
      start_date,
      end_date,
      agent_id
    });

    res.locals.createdId = policy.policy_id;
    return res.status(201).json({ success: true, data: policy });
  } catch (err) {
    next(err);
  }
};

const getRenewals = async (req, res, next) => {
  try {
    const { role_name, user_id } = req.user;
    const agentId = role_name === 'Sales Agent' ? user_id : (req.query.agent_id ? parseInt(req.query.agent_id) : null);

    const renewals = await Policy.findRenewals(agentId);
    return res.status(200).json({ success: true, data: renewals });
  } catch (err) {
    next(err);
  }
};

const updatePolicy = async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
    }

    const id = parseInt(req.params.id);
    const existing = await Policy.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'Policy not found' });
    }

    const updated = await Policy.update(id, req.body);
    return res.status(200).json({ success: true, data: updated });
  } catch (err) {
    next(err);
  }
};

const deletePolicy = async (req, res, next) => {
  try {
    const id = parseInt(req.params.id);
    const existing = await Policy.findById(id);
    if (!existing) {
      return res.status(404).json({ success: false, message: 'Policy not found' });
    }

    await Policy.remove(id);
    return res.status(200).json({ success: true, message: 'Policy deleted successfully' });
  } catch (err) {
    next(err);
  }
};

module.exports = { getPolicies, createPolicy, getRenewals, updatePolicy, deletePolicy };

File: dotnet-api/Controllers/PoliciesController.cs
using ActivityTrackerAPI.Helpers;
using ActivityTrackerAPI.Models.DTOs;
using ActivityTrackerAPI.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace ActivityTrackerAPI.Controllers;

/// <summary>Insurance policy management endpoints</summary>
[ApiController]
[Route("api/policies")]
//[Authorize]
[Produces("application/json")]
public class PoliciesController : ControllerBase
{
    private readonly IPolicyService _policyService;

    public PoliciesController(IPolicyService policyService)
    {
        _policyService = policyService;
    }

    /// <summary>List policies with optional filters and pagination</summary>
    [HttpGet]
    [ProducesResponseType(200)]
    public async Task<IActionResult> GetPolicies(
        [FromQuery] uint? product_type_id,
        [FromQuery] uint? agent_id,
        [FromQuery] string? date_from,
        [FromQuery] string? date_to,
        [FromQuery] string? search,
        [FromQuery] int page = 1,
        [FromQuery] int limit = 20)
    {
        var roleName = User.GetRoleName();
        var userId = User.GetUserId();

        uint? filterAgentId = agent_id;
        if (roleName == "Sales Agent")
            filterAgentId = userId;

        limit = Math.Min(limit, 100);
        var (data, total) = await _policyService.GetAllAsync(
            product_type_id, filterAgentId, date_from, date_to, search, page, limit);

        return Ok(new
        {
            success = true,
            data,
            pagination = new { page, limit, total, totalPages = (int)Math.Ceiling((double)total / limit) }
        });
    }

    /// <summary>Create a new policy</summary>
    [HttpPost]
    [ProducesResponseType(201)]
    [ProducesResponseType(400)]
    public async Task<IActionResult> CreatePolicy([FromBody] CreatePolicyRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var policy = await _policyService.CreateAsync(
            request.CustomerName, request.PolicyNumber, request.LeadId,
            request.ProductTypeId, request.Premium,
            request.StartDate, request.EndDate, request.AgentId);

        return StatusCode(201, new { success = true, data = policy });
    }

    /// <summary>Get policies expiring in 30/60/90 days</summary>
    [HttpGet("renewals")]
    [ProducesResponseType(200)]
    public async Task<IActionResult> GetRenewals([FromQuery] uint? agent_id)
    {
        var roleName = User.GetRoleName();
        var userId = User.GetUserId();

        uint? filterAgentId = agent_id;
        if (roleName == "Sales Agent")
            filterAgentId = userId;

        var data = await _policyService.GetRenewalsAsync(filterAgentId);
        return Ok(new { success = true, data });
    }

    /// <summary>Update a policy</summary>
    [HttpPut("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> UpdatePolicy(uint id, [FromBody] UpdatePolicyRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, message = "Validation failed", errors = ModelState });

        var existing = await _policyService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "Policy not found" });

        var updated = await _policyService.UpdateAsync(id,
            request.CustomerName, request.PolicyNumber, request.LeadId,
            request.ProductTypeId, request.Premium, request.StartDate, request.EndDate,
            request.AgentId, request.RenewalNotified30, request.RenewalNotified60, request.RenewalNotified90);

        return Ok(new { success = true, data = updated });
    }

    /// <summary>Delete a policy (Admin only)</summary>
    [HttpDelete("{id:int}")]
    [ProducesResponseType(200)]
    [ProducesResponseType(403)]
    [ProducesResponseType(404)]
    public async Task<IActionResult> DeletePolicy(uint id)
    {
        var roleName = User.GetRoleName();
        if (roleName != "Admin")
            return StatusCode(403, new { success = false, message = "Access denied. Required roles: Admin" });

        var existing = await _policyService.GetByIdAsync(id);
        if (existing == null)
            return NotFound(new { success = false, message = "Policy not found" });

        await _policyService.DeleteAsync(id);
        return Ok(new { success = true, message = "Policy deleted successfully" });
    }
}
